### DISCUSSION TITLE
[Enhancement] MAUI Should be what UNO is trying to do

### DISCUSSION THREAD

**Discussion by jhancock4d** [dotnet/maui#558] [Ideas]
## Summary
Maui seems to be just a warmed over version of Xamarin Forms.  Microsoft needs to go big and bold and make MAUI into the truly cross platform competitor it needs to be, and the obvious approach is to buy UNO and then make MAUI have all of the promised features (single project stuff) but have it working across Web, Windows, Android, iOS, MacOS, Linux, and fully implement WinUI 3 with pixel perfect rendering across platforms.

And, while you're at it, create a conversion tool that takes WinForms, WPF and UWP apps and converts them to this new project format and substitutes controls etc. automatically and provides todos for anything that can't be automatic with clear instructions on what needs to be done.  

Build the entire thing open source and let other people/companies work on the conversion tool as well.

If you did this, you'd breathe new life into Windows development, unify windows development so that Windows developers can reach all the other platforms AND truly build on one Win UI 3 centric platform, and you'd have a true competitor to flutter.  Uno has done most of the hard work here.  The delta of full Win UI 3 support isn't that large anymore, and then you'd need to focus on getting all of the base plugins working on all platforms (sensors, bluetooth, web view, video player, openidconect view, etc.) Again, most of this work is done for almost everything other than web because of the work of Xamarin Essentials.

If you showed an early alpha at Build, you'd generate such excitement with this clear, obvious, and consistent vision that you'd have major developer interest, especially with Xamarin allowing developers to write C# code against the native apis so you don't need objective c or java or kotlin or C++ developers, which flutter still requires to get plugins and custom stuff working. And if you had Telerik and DevExpress etc. on stage with you telling the world that they're going to support this single surface area and that they'd work with the converter people to convert their stuff to Win UI 3 controls as well, you'd have a serious win.

And because Xamarin does the remote iOS simulator, you'd win against flutter even more. 

And because C# isn't awful like Dart (very immature language, eventually it might not be awful but it is massively limiting now), you'd win there too especially with AOT and tree shaking getting you close to Flutter in size, especially if you allowed chunking on the web to chunk pages like Angular or Blazor.

Demo with code as UI like flutter/jetpack/Swift UI while you're at it, but show that XAML also works and that the converter will convert XAML to Code UI if you wish or keep XAML and that you can mix and match in the same project and you'd have a huge win, because then developers won't need to use XAML if they don't like it. (lots of people do not)

And what that does, is mean that Microsoft doesn't have to be building Blazor, Xamarin, Win UI 3, UWP, WPF, and WinForms and instead can focus on this single UI layer just like .net core allowed focusing on "one .net" and you're working to eliminate the mono as a separate project and bring it into one .net so no duplication.

And with the proper buyin, Microsoft as a whole would become more productive and agile because you'd have a single development set for all of your other apps and it would be vastly more effective and memory friendly than the current React Native insanity that's going on in Microsoft if done correctly. (and if you built this to be extensible, React Native could render using the same framework as well while you were switching and it would mean that Microsoft apps could share the rendering and look identical across all platforms. Think Outlook which is being redone in react native or Electron, instead being migrated over over time using the single rendering surface of this)

## Intended Use Case
All development on all platforms including windows going forward under Microsoft.

**Comment by UltraBurstXD** [comment]
I don't like that they are ignoring Linux, stating that MAUI support is under community hands. But Tizen, an OS that nearly anyone doesn't know/care has slight priority! 

Community doesn't have great resources like many think... Many Linux Desktop are struggling to keep things working... It's really Armageddon battleground. Job is endless. There are many bugs to be caught and destroyed, many things to be improved... When will they have spare time to play with MAUI?! And from what I know, many Linux devs (Even I) don't care too much for C# (apart from some Game Engines)... C# isn't commonly used there.

The only strong side of Linux OS is Server and Cloud, because there may be great payoff chance. Desktop depends mainly from donations, some companies got their way out with pay-what-you-want model or Linux Server.

**Reply by GeraudFabien** [reply]
Tizen is support by Samsung as linux is supported by Community. So as far i know for the team point of view i assume it's the exact same state.

Web support is more important than linux in my point of view since if you support web you support linux through the browser... Witch is better than no support.

**Reply by UltraBurstXD** [reply]
I understand your point of view, but Linux Community has scarcity of resources issue, and whether web or cloud is more important I can't say nothing. We all know the limitations of (great :expressionless:) Web or Cloud, besides the pricing issues to get things working on hosting platforms.

We are all struggling to maintain and improve ours distros... It's very hard... It's unlikely someone from Community will step out and help... And BTW, does Apple care for MAUI or C#?

**Reply by saint4eva** [reply]
Users care

**Reply by UltraBurstXD** [reply]
Some minority we could say... C# is more widely used on Windows, like Swift and Objective-C are most used on Apple systems...

People were getting exciting and started seeing C# as another cross-platform solution, but things aren't going nowhere as we can see. C# had great potential to destroy Java for once and for all in the start.

**Reply by JohnGalt1717** [reply]
The point, I think, is kind of that Microsoft needs to make Xamarin/WinUI 3 into a Flutter competitor and be cross platform for C# to flourish instead of continually losing market share as it is now.

That means pixel perfect rendering across platforms using the Win UI 3 system like Uno is doing including Web, Windows, iOS, Android, Mac and Linux.

Given all of the censorship on Android and iOS it's no longer OK to design your apps for iOS and Android and ignore web as a backstop with full functionality and having to dev at least twice (Xamarin + whatever) isn't a viable solution when you can do it all in Flutter with Dart and the results are already very good and continually getting better. (no it isn't perfect)

The lack of a cohesive strategy by Microsoft to combat and be better than Flutter at what Flutter does, which is clearly virtually the only use case for Xamarin/MAUI, is deeply concerning and shows a continued lack of direction for C# on platforms that needs to be addressed by Microsoft.  

I'd argue, in fact, that Microsoft needs to create a seamless upgrade path for at least WPF, UWP and Xamarin to an integrated cohesive WinUI3-based cross platform development tool that works best and natively on Windows and drop absolutely all other support for everything else other than as continued legacy support. Obviously, this needs to happen over time, and the tooling needs to be created to make the conversion happen, but with help from the 3rd party control manufacturers, upgrading WPF/Xamarin/UWP XAML to Winui3 xaml is doable as is translating the code.   If Microsoft did this in the open with the direct pathway for people to take their existing apps and easily migrate them to this singular platform (that would look very much like Win UI 3 desktop apps but with all of the other platforms, which is just WPF ++) then we'd have confidence back to develop on  Windows and for windows. Until then, Windows is just a shell to Office + A Web Browser for the vast majority of people which is a dead end approach to Windows because Linux is just as good a shell to a web browser as Windows and costs WAY less money.

**Reply by UltraBurstXD** [reply]
I don't think Flutter is reliable solution for everything. It can be slower than **Java** in many scenarios, among others annoying things, but it's shaping out nearly fine with many contributors. It's strong point can be android and web from what I can see, iOS support can be laughable. Desktop I can't say much, we all know it's still in beta.

**Reply by UltraBurstXD** [reply]
I don't mind Java, Sun Microsystems was trying to bring cross platform solution "Write once, run everywhere". In reality we know that was all lie. One have to rewrite some or nearly all parts of program for each platform. C# can be more faster than Dart. In my benchmarks, .NET5 is more faster than Java 16 in many cases and sometimes uses less resources. 

I don't think we should compare languages by age, Rust while not really "cross platform" is the only true language that surpassed C/C++ in nearly everything: speed, memory safety, etc., It only lack a reliable GUI framework. Rust being accepted in many game development environments and web now.

I don't like very well VM based languages and GCs, but C# is little different. But it lacks direction and great plan. It should teach Java and Dart who's the boss.

**Reply by LazyCoderSuhe** [reply]
英文水平有限,就来中文.感觉.多家公司的做跨平台中 
,需求: 
1 , 成本 ;
2,UI 一致性;
3,入门剪单;,
4 开源.
5 有技术保障.
基于这些 MAUI的发展的 方向是
1 xaml 标准化
2 MAUI - WInUI 设计之初就应该目标一致,否者就分开
3 UI 控件 的统一效果呈现.

**Comment by albertomandlate** [comment]
> I don't like that they are ignoring Linux, stating that MAUI support is under community hands. But Tizen, an OS that nearly anyone doesn't know/care has slight priority! 
> 
> 
> 
> Community doesn't have great resources like many think... Many Linux Desktop are struggling to keep things working... It's really Armageddon battleground. Job is endless. There are many bugs to be caught and destroyed, many things to be improved... When will they have spare time to play with MAUI?! And from what I know, many Linux devs (Even I) don't care too much for C# (apart from some Game Engines)... C# isn't commonly used there.
> 
> 
> 
> The only strong side of Linux OS is Server and Cloud, because there may be great payoff chance. Desktop depends mainly from donations, some companies got their way out with pay-what-you-want model or Linux Server.
> 
> 

Linux is not being ignored. First things first. Having a single project, infuture, will be very fluid for developers to upgrade to a new OS on a GA of a new framework/maui.

But yes, I think that Microsoft needs to run and tell everyone about their plans to the futuro (I think that they are not so dar of that).

**Reply by UltraBurstXD** [reply]
Well, if Apple cares for C#, than many enthusiasts Linux users, I don't know... C# may be very good for game scripting. If they port most important legacy Windows UI frameworks to others platforms, we could benefit from many Game Engines strictly written for Windows. And maybe some interesting apps...

We could have plenty apps available... I really don't understand the priority (OK, sure we have iOS and macOS that are indispensable), many devs care more about Apple. Apple, in the end doesn't care about ANYONE. We all know that...

**Comment by rperazzolli** [comment]
It would be very important to use the same tools on phones, personal computers and web, in a similar way of UNO platform, but with support of Microsoft.

**Comment by akhanalcs** [comment]
Totally agree. 
.NET MAUI feels like an effort that's too little too late against modern tools the competition offers. 
I hope MAUI team takes this suggestion seriously.

**Comment by AslanAmca** [comment]
I largely agree, but it's too late to say. Unfortunately, Microsoft will not do such a thing. I hope MAUI will be successful, but I don't think it will.

There are 2 things Microsoft will do in the near future; UNO for User Interface and Duende Software (Identity Server) for Authorization must be purchased.

**Reply by HannsKanns** [reply]
Has anyone in the world ever done a project with over 2,000,000 lines of code using Xamarin or Uno? I don't think so. So I know what I am talking about. Neither Xamarin nor Uno are capable of handling such large projects reliably on all mobile platforms. I have invested 2 full years of manpower and have ALWAYS run into terrible limitations. Since I need to stick with C# and the .NET Platform for various reasons, my hopes lie with Maui. Finally Microsoft has committed initiative and probably money to provide a reasonable and reliable mobile development system. I believe it will be good. If anyone in the world should have successfully handled such a big project as mine with Xamarin or Uno, they are welcome to contact me. I don't think there will be anyone.

**Comment by EgorSozonov** [comment]
Totally agree with the original post. If only Microsoft listened to you, they would be in a much better position in the future. Unfortunately, they are not known for good software choices (anyone remember ActiveX? what about Silverlight? UWP? Windows Phones?), with .NET CLR being a very striking difference. 
.NET is a really, really great platform. As in, better than the JVM, and more deserving of the world dominance than Java. But being great, the core software running on it is ruined by Microsoft's faulty, unstable choices. Remember how many fickle and backward-incompatible changes ASP.NET has had (I'm sick of running through documentation from old versions which is very alike but subtly different from the latest version). What about WinForms? WPF? And Xamarin.Forms which is now superseded by "MAUI"? Yet Xamarin.Forms was everyone's favorite beat-up bag, everyone hated it, and since MAUI is just an evolution of more of the same, who's to say it also won't need replacing several years in the future?
Personally, as a C# programmer all my career, I've recently spent some time analyzing my options for getting into mobile and made a solid choice to switch to Dart. Yes, it's got a weak, immature platform and a separate language. Yes, it'll take years for Dart VM to get on any semblance of same level as .NET. But it's got Flutter, the one and only true approach to building multi-platform UIs. And that's what matters, that's what's responsible for Dart's explosive popularity growth seemingly out of nowhere.

**Comment by Wynadorn** [comment]
To be honest, I have no idea what MAUI is trying to achieve in terms of UI styling. I assume apps targeting windows will have a Fluent Design implemented through WinUI 3.

But what about Android and iOS? Will my controls be rendered in Material & Cupertino/iOS designs? Will it be Fluent Design? Will it be barebones, aka DIY?

Hope it's not too off topic, but if anyone knows, please let me know.

**Edit:**
For me Maui Blazor was **the** solution.
Web was so many UI-control libraries available which you can use to build your front-end with.
I'm currently using the Ionic framework.

**Reply by Wynadorn** [reply]
In February David (@davidortinau) from the Xamarin team wrote the following:

> **Design**
> The design requirements for applications that target multiple platforms have changed over the years. In the early days of Xamarin.Forms, most customers wanted their Android app to look very Android, and iOS app to look very iOS. We have introduced in Xamarin.Forms more and more styling and controls that look the same across platforms, and added the Visual control API along with Material Design components for iOS to match what you get on Android. These efforts went part of the way towards satisfying what customers have shown us they require, and in .NET MAUI we are laying the foundation to go even further.

https://devblogs.microsoft.com/xamarin/the-new-net-multi-platform-app-ui-maui/

>We are enabling full-app themes through styling for Fluent UI and Material Design, filling out the existing controls with the necessary styling APIs and visual states, and adding support for control templates as we are able (see RadioButton for a recent example of this).
>
>On the experiment front, we also have kicked off a new Visual strategy that uses a combination of native controls and drawn controls. We will have more information to share about that later this year. For customers needing the 1:1 design across platforms and looking for even better performance, but also having a requirement for native platform support, we think this will be on target.

It seems very similar to what the Uno Platform is doing. But again, no idea about the current status.

**Reply by Wynadorn** [reply]
This seems promising 👀https://github.com/dotnet/Microsoft.Maui.Graphics.Controls

**Reply by saint4eva** [reply]
Maui targets the various platforms with their own idiomatic design languages. However, there is an effort for drawn controls that would make it possible to have the same look and feel across the targeted platform - though the effort is still experimental. E.g. Having your apps with Fluent Design look and feel across Windows, Android, iOS and Tizen.
